samza-core/src/main/java/org/apache/samza/clustermanager/StandbyContainerManager.java

```diff
@@ -0,0 +1,459 @@
+      // This resource cannot be used to launch this standby container, so we make a new anyhost request
+      log.info("Running standby container {} on host {} does not meet standby constraints, cancelling resource request, releasing resource, and making a new ANY_HOST request",
+          containerID, samzaResource.getHost());
+      resourceRequestState.releaseUnstartableContainer(samzaResource, preferredHost);
+      resourceRequestState.cancelResourceRequest(request);
+      containerAllocator.requestResource(containerID, ResourceRequestState.ANY_HOST);
+      samzaApplicationState.failedStandbyAllocations.incrementAndGet();
+    } else {
+      // This resource cannot be used to launch this active container container, so we initiate a failover
+      log.warn("Running active container {} on host {} does not meet standby constraints, cancelling resource request, releasing resource",
+          containerID, samzaResource.getHost());
+      resourceRequestState.releaseUnstartableContainer(samzaResource, preferredHost);
+      resourceRequestState.cancelResourceRequest(request);
+
+      // if this active-container has never failed, then simple request anyhost
+      if (!samzaApplicationState.failedContainersStatus.containsKey(containerID)) {
+        log.info("Requesting ANY_HOST for active container {}", containerID);
+        containerAllocator.requestResource(containerID, ResourceRequestState.ANY_HOST);
+      } else {
+        log.info("Initiating failover for active container {}", containerID);
+        // we use the activeContainer's last known resourceID to initiate the failover
+        String lastKnownResourceID = samzaApplicationState.failedContainersStatus.get(containerID).getResourceID();
+        initiateActiveContainerFailover(containerID, lastKnownResourceID, containerAllocator);
+      }
+
+      samzaApplicationState.failedStandbyAllocations.incrementAndGet();
+    }
+  }
+
+  public void handleExpiredResourceRequest(String containerID, SamzaResourceRequest request,
+      Optional<SamzaResource> alternativeResource, AbstractContainerAllocator containerAllocator,
+      ResourceRequestState resourceRequestState) {
+
+    if (StandbyTaskUtil.isStandbyContainer(containerID) && alternativeResource.isPresent()) {
+      // A standby container can be started on the anyhost-alternative-resource rightaway provided it passes all the
+      // standby constraints
+
+      log.info("Handling expired request, standby container {} can be started on alternative resource {}", containerID, alternativeResource.get());
+
+      checkStandbyConstraintsAndRunStreamProcessor(request, ResourceRequestState.ANY_HOST, alternativeResource.get(),
+          containerAllocator, resourceRequestState);
+
+    } else if (StandbyTaskUtil.isStandbyContainer(containerID) && !alternativeResource.isPresent()) {
+      // If there is no alternative-resource for the standby container we make a new anyhost request
```

Comment:
The logic of checking for empty resources feels a bit mis-placed in this class.

it introduces multiple multiple branches in this class and in the HostAwareAllocator as well :)  is there any room to reduce the complexity?

Reply:
Let me try separating it out by container-type (standby v. active) : handleExpiredRequestForActive and handleExpiredRequestForStandby. 
That would help i think.

This overall method is basically saying a resource request has expired, and optionally an alternate anyHost resource is available.